services/src/main/java/org/keycloak/protocol/oid4vc/issuance/credentialbuilder/SdJwtCredentialBuilderFactory.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2024 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.protocol.oid4vc.issuance.credentialbuilder;
+
+import org.keycloak.component.ComponentModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.protocol.oid4vc.issuance.VCIssuerException;
+import org.keycloak.protocol.oid4vc.model.Format;
+import org.keycloak.provider.ProviderConfigProperty;
+
+import java.util.List;
+import java.util.Optional;
+
+import static org.keycloak.protocol.oid4vc.issuance.signing.VCSigningServiceProviderFactory.ISSUER_DID_REALM_ATTRIBUTE_KEY;
+
+/**
+ * @author <a href="mailto:Ingrid.Kamga@adorsys.com">Ingrid Kamga</a>
+ */
+public class SdJwtCredentialBuilderFactory implements CredentialBuilderFactory {
+
+    @Override
+    public String getId() {
+        return Format.SD_JWT_VC;
+    }
+
+    @Override
+    public String getHelpText() {
+        return "Builds verifiable credentials on the SD-JWT format (https://drafts.oauth.net/oauth-sd-jwt-vc/draft-ietf-oauth-sd-jwt-vc.html).";
+    }
+
+    @Override
+    public List<ProviderConfigProperty> getConfigProperties() {
+        return null;
+    }
+
+    @Override
+    public CredentialBuilder create(KeycloakSession session, ComponentModel model) {
+        RealmModel realm = session.getContext().getRealm();
+        String issuerDid = Optional.ofNullable(realm.getAttribute(ISSUER_DID_REALM_ATTRIBUTE_KEY))
+                .orElseThrow(() -> new VCIssuerException("No issuerDid configured."));
```

Comment:
@Captain-P-Goldfish I am not very familiar with DID, but your proposal seems reasonable.
Tagging @babisRoutis here, as he is one of my go-to experts on DID.

Reply:
@Captain-P-Goldfish While I agree with the realm-uri as fallback, I dont fully understand the comment about ```VerifiableCredentials```.  What other credential-types are we able to issue here?

Reply:
For what is worth, and given that we dealing with SD-JWT-VC format: 

The `iss` claim must be a URI (https://www.ietf.org/archive/id/draft-ietf-oauth-sd-jwt-vc-06.html#section-3.2.2.2)
Also, taking into account [Issuer-signed JWT Verification Key Validation](https://www.ietf.org/archive/id/draft-ietf-oauth-sd-jwt-vc-06.html#name-issuer-signed-jwt-verificat), `iss` must be an HTTPS URI.

Please note that DID resolution for issuer's pub (signing) key has been removed from SD-JWT-VC spec. This means that DID resolution is a perfectly valid option but it is not included in the SD-JWT-VC spec. 

So, I think that a builder that builds SdJwtVc credentials should take into account how the issuer wants the wallets (or verifiers) to discover his pub signing key. 

Two options are available:
- `iss` to be an HTTPS URI, that can be use with `/.well-known/jwt-vc-issuer/` to get pub singing key
- `iss` to be an HTTPS URI combined with an `x5c` header element.

Reply:
@babisRoutis As far as I understand, the spec you linked(https://www.ietf.org/archive/id/draft-ietf-oauth-sd-jwt-vc-06.html#name-issuer-signed-jwt-verificat) does not enforce an https-uri, but just explains what to do in case it is an https-uri(see ``` and if the iss value contains an HTTPS URI```). Thus it does not need to be an https-uri, but could very well be a did.

Reply:
Hi @wistefan Good point, but I guess it is a matter of interpretation :smile: 

SD-JWT-VC requires (from issuer, wallet, verifier) to support one of the two methods described in the above paragraph. In both cases `iss` is HTTPS URI.

Spec gives room for other cases (like DID), yet clearly states that it is outside the spec's scope.

>Separate specifications or ecosystem regulations MAY define rules complementing the rules defined above, but such rules are out of scope of this specification

So, I think that the builder, in question, should allow the caller to define primarily what is supported by the spec.

Reply:
> @Captain-P-Goldfish While I agree with the realm-uri as fallback, I dont fully understand the comment about `VerifiableCredentials`. What other credential-types are we able to issue here?

In the FUNKEN challenge we learned that a credential must not necessarily be a `VerifiableCredential`. Let us assume we want to support additional formats like simple `json` no `vc` in it. In that case DID's become obsolete. This would also apply if `VerifiableCredentials` do not gain the acceptance in the community they are expected to get. In that case it would be also very likely that another format is being developed.
I just want to point out that we should try to keep everything open as much as possible and that we prevent to build any blockades. If we stiffen too much on VerifiableCredentials it will also be difficult to add custom-formats which several companies will surely try to implement.

Reply:
> > Separate specifications or ecosystem regulations MAY define rules complementing the rules defined above, but such rules are out of scope of this specification
> 
> So, I think that the builder, in question, should allow the caller to define primarily what is supported by the spec.

In that case we need a configuration option for the issuer. I could also imagine that it could be interesting to use a realm-level configuration value as default and use credential-format specific iss-values as well, but we can maybe wait with this until it is actually requested. 
But making the `iss` value configurable and not calling it did is the main issue in my eyes.
And I also think that it would be great if we could avoid to pass such parameters through more than two methods. If the chain goes longer it would be better to read it directly from the configuration where it is needed. Makes the code easier to read.

Reply:
> But making the `iss` value configurable and not calling it did is the main issue in my eyes.

:+1: 

If I could make a proposal:
- OpenId4VCI defines the [`CredentialIssuer`](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#name-credential-issuer-metadata) which is used to advertise and discover OpenId4VCI Credential Issuer Metadata.
- SD-JWT-VC spec uses the term `issuer` which can be used to populate the `iss` claim and may also be used to advertise SD-JWT-VC issuer metadata.

Those two can be combined. A builder for SdJwtVc could use the value of the `CredentialIssuer` as the default value of `iss` claim, unless it is overridden (for example to support DID, in the context of a specific use case).  Configurations options should be present for  populating `kid` ([Recommended](https://www.ietf.org/archive/id/draft-ietf-oauth-sd-jwt-vc-06.html#name-jwt-vc-issuer-metadata-resp) by sd-jwt-vc) and/or `x5c`.

Reply:
To add to the discussion, actually for one of our applications at adorsys, we are simply misusing the `issuerDid` realm attribute with an HTTPS URI value. Prior to making the issuer claim overridable with credential-format specific configurations, would you agree with a new `issuerUri` realm attribute taking precedence over `issuerDid` as the issuer claim value when building SD-JWTs?

Reply:
> Configurations options should be present for populating kid ([Recommended](https://www.ietf.org/archive/id/draft-ietf-oauth-sd-jwt-vc-06.html#name-jwt-vc-issuer-metadata-resp) by sd-jwt-vc) and/or x5c.

The `kid` header, at least, is currently configurable via signing service properties to my knowledge.

https://github.com/keycloak/keycloak/blob/1d7f07fc53c78a7cf7699ac0169ca42b31722413/services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningServiceProviderFactory.java#L52

Reply:
> To add to the discussion, actually for one of our applications at adorsys, we are simply misusing the `issuerDid` realm attribute with an HTTPS URI value. Prior to making the issuer claim overridable with credential-format specific configurations, would you agree with a new `issuerUri` realm attribute taking precedence over `issuerDid` as the issuer claim value when building SD-JWTs?

I would suggest to completely exchange the DID configuration attribute for an `credential-iss` attribute which itself can be a DID. Why separating it? If we separate it I would rather suggest following my other proposal that we allow `iss` values to be configured for credential specific formats like `jwt_vc` but I have actually no good concept how to configure this appropriately yet

Reply:
> I would suggest to completely exchange the DID configuration attribute for an `credential-iss` attribute which itself can be a DID. Why separating it? If we separate it I would rather suggest following my other proposal that we allow `iss` values to be configured for credential specific formats like `jwt_vc` but I have actually no good concept how to configure this appropriately yet

I am concerned by the fact that with a unified configuration attribute, we may not be able to concurrently support issuing credential formats that would preferably go with one or the other issuer claim type. To address this point, can't we just proceed with your other proposal right away? That is, a default `credentialIss` attribute at the realm (there seems to be a preference for camelCase), with the possibility to override the value with format-specific credential builder configurations? I do not think this will not survive a deeper understanding of the `jwt_vc`  format.

Reply:
Thank you @babisRoutis for providing more details.

> At the realm level, there should be the property `CredentialIssuer` (json attribute name=`credential_issuer`) as it is defined in the OpenId4VCI specification. Probably that's already the case.

Oh you're right... There is an existing method to get that value.

https://github.com/keycloak/keycloak/blob/a4d70ad6d2ac76a2711b5b975bb3d75770fc650a/services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerWellKnownProvider.java#L127-L135

Reply:
Thanks for pointing out the necessary topics @babisRoutis, can't argue with that :-)
With this it looks good to me.

Reply:
Following this discussion, I updated the PR to integrate the ideas expressed, mainly to use the CredentialIssuer already exposed on the metadata endpoint as the default (or fallback) `iss` claim when building SD-JWT VCs. Thank you all for your participation.